CONTRIBUTING.md

```diff
@@ -1,69 +1,138 @@
+Before reporting an issue on the [issue tracker](https://github.com/tui-rs-revival/ratatui/issues),
+please check that it has not already been reported by searching for some related keywords. Please
+also check [tui-rs issues](https://github.com/fdehau/tui-rs/issues/) and link any related issues
+found.
 
-- Recreate the issue
-- Start by referencing the **original issue**: ```Referencing issue #[<issue number>](<original issue link>)```
-- Then, paste the original issues **opening** text
+## Pull requests
 
-You can then resume the conversation by replying to this new issue you have created.
+All contributions are obviously welcome. Please include as many details as possible in your PR
+description to help the reviewer (follow the provided template). Make sure to highlight changes
+which may need additional attention or you are uncertain about. Any idea with a large scale impact
+on the crate or its users should ideally be discussed in a "Feature Request" issue beforehand.
+
+### Keep PRs small, intentional and focused
+
+Try to do one pull request per change. The time taken to review a PR grows exponential with the size
+of the change. Small focused PRs will generally be much more faster to approved and merge. Don't
+include large refactoring or formatting changes in PRs
+
+### Search tui-rs for similar work
 
-### Closing Issues
+Tui-rs has a large amount of history in issues, discussions and PRs that is relevant to this fork.
+Please make sure to search and link any previous issues / discussions related to your PR.
 
-If you close an issue that you have "imported" to this fork, please make sure that you add the issue to the **CLOSED_ISSUES.md**. This will enable us to keep track of which issues have been closed from the original repo, in case we are able to have the original repository transferred.
+### Use conventional commits
 
-# Contributing
+We use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) and check for them as
+a lint build step. To help adhere to the format, we recommend to install
+[Commitizen](https://commitizen-tools.github.io/commitizen/). By using this tool you automatically
+follow the configuration defined in [.cz.toml](.cz.toml). Your commit messages should have enough
+information to help someone reading the [CHANGELOG](./CHANGELOG.md) understand what is new just from
+the title. The summary helps expand on that to provide information that helps provide more context,
+describes the nature of the problem that the commit is solving and any unintuitive effects of the
+change. It's rare that code changes can easily communicate intent, so make sure this is clearly
+documented here.
+
+When the PR is close to approval, please ensure that it is rebased against main, squashed into an
+appropriate amount of commits (usually 1). We don't need to have a commit for every small change
+that was made during the PR process. Use `git push --force` to update the branch associated with
```

Comment:
This needs context. What can you do with force push? Probably
- Amend commits incrementally if they do not need to be separate (`--amend --no-edit`)
- Squash commits, or otherwise edit them (reword etc.)

Reply:
I probably need to link to some other doc here. But yep, context would make sense here (iow, why am I asking for this)

Reply:
I don't think we should be going this in depth on how to do any of those git operations, seems out of scope for a contributing guide. Additionally, we haven't ever done anything else than squash and merge so even if we are faced with a large number of commits they always get squashed, so imho I think we should drop everything after the first sentence. It would also help to not overcomplicate the contributing guide.

Reply:
The problem I'm trying to solve here is a communication one. We use commit messages as our change log, which means that a commit message should be written with the library user in mind. It's the best place to capture the intent of the change and a way of communicating the change to someone so that they don't have to read the entire source code to work out how something works or find the changes docs.

When we accept PRs that have a bunch of fixes to little things found as part of the review it's just as important to communicate the end state as it was when we started. But things like "fixed comments", "cargo fmt", etc. are noise that distracts from that.

We don't necessarily need to document the steps in the guide, but perhaps instead something like try to aim for a single commit that has a commit message that documents your new feature (and then linking to some other docs).

It's my experience that some of the weirdest annoying bugs or legacy "can't change that because we don't know why it's there" things are strongly associated with messy commits where care was not taken to remove these sorts of things.

Anyway, I'll reword this to try to make some of that clear and find some good external docs to link to so our doc is smaller.